Add pagination support to ReferencesClient.

gdziadkiewicz opened the pull request:
Fixes #1664 

- [x] Remove convention test exclusion

- [x] Add `ApiOptions` overload
  - make existing methods call new overload with `ApiOptions.None`

- [x] Update unit tests
  - Adjust for `ApiOptions` parameter
  - Add `null`/`""` checks for `ApiOptions` overloads

- [x] Add pagination Integration tests
  - `ReturnsCorrectCountOfXWithoutStart`
  - `ReturnsCorrectCountOfXWithStart`
  - `ReturnsDistinctXBasedOnStartPage`

- [x] Decide what to do with `GetAllForSubNamespace` endpoint

ryangribble commented on Octokit.Tests/Clients/ReferencesClientTests.cs:
```diff
@@ -106,8 +110,10 @@ public class TheGetAllForSubNamespaceMethod
+                await Assert.ThrowsAsync<ArgumentNullException>(() => client.GetAllForSubNamespace(1, "subnamespace", null));
 
                 await Assert.ThrowsAsync<ArgumentException>(() => client.GetAllForSubNamespace("", "name", "heads"));
                 await Assert.ThrowsAsync<ArgumentException>(() => client.GetAllForSubNamespace("owner", "", "heads"));
```
we also need empty string test cases for the 4 parameter overload (and below for the 3 parameter one `1, "", options)`)

gdziadkiewicz commented:
@ryangribble I will work on it today. I was low on spare time in past few days, sorry for the delay.

ryangribble commented:
No apology necessary,  we all do this in our spare time when life permits 😝 

just wanted to make sure you weren't stuck

gdziadkiewicz commented:
@ryangribble
I have added the missing unit tests, added and ran integration tests, but there is a problem - `GetAllForSubNamespace` endpoint ignores requested pagination. Do you have any idea how to handle that?

ryangribble commented:
I've raised a ticket with GitHub support about the API pagination not being honoured for the subnamespace call

Repro:

✅  https://api.github.com/repos/octokit/octokit.net/git/refs?per_page=1

💀 https://api.github.com/repos/octokit/octokit.net/git/refs/tags?per_page=1

ryangribble commented on Octokit.Tests.Integration/Clients/ReferencesClientTests.cs:
```diff
@@ -63,6 +63,61 @@ public ReferencesClientTests()
+        var firstRefsPage = await _fixture.GetAll("octokit", "octokit.net", startOptions);
+        var secondRefsPage = await _fixture.GetAll("octokit", "octokit.net", skipStartOptions);
+
+        foreach (var r1 in firstRefsPage)
```
The nested loops are a bit clunky, is there a more succinct way to express this?  

eg `Assert.False(firstRefsPage.Any(x => secondRefsPage.Contains(x)));`

gdziadkiewicz replied:
Changed it.

ryangribble commented:
To get the travis build passing, pull in latest master to your branch :+1:

In terms of the pagination issue up stream, GitHub support confirmed it's a problem although it looks like it still isnt fixed.  I think it should be fine to leave it implemented in octokit as is, since hopefully they will fix it eventually 😁

ryangribble commented on Octokit.Tests.Integration/Clients/ReferencesClientTests.cs:
```diff
@@ -63,27 +63,243 @@ public ReferencesClientTests()
+        var firstRefsPage = await _fixture.GetAll(7528679, startOptions);
+        var secondRefsPage = await _fixture.GetAll(7528679, skipStartOptions);
+
+        foreach (var r1 in firstRefsPage)
```
another assertion to refactor here

gdziadkiewicz replied:
Done.

ryangribble commented on Octokit.Tests.Integration/Clients/ReferencesClientTests.cs:
```diff
@@ -63,27 +63,243 @@ public ReferencesClientTests()
+        var firstRefsPage = await _fixture.GetAllForSubNamespace("octokit", "octokit.net", "heads", startOptions);
+        var secondRefsPage = await _fixture.GetAllForSubNamespace("octokit", "octokit.net", "heads", skipStartOptions);
+
+        foreach (var r1 in firstRefsPage)
```
and here

gdziadkiewicz replied:
Done.

ryangribble commented on Octokit.Tests.Integration/Clients/ReferencesClientTests.cs:
```diff
@@ -63,27 +63,243 @@ public ReferencesClientTests()
+        var firstRefsPage = await _fixture.GetAllForSubNamespace(7528679, "heads", startOptions);
+        var secondRefsPage = await _fixture.GetAllForSubNamespace(7528679, "heads", skipStartOptions);
+
+        foreach (var r1 in firstRefsPage)
```
and here

gdziadkiewicz replied:
Done.

ryangribble commented:
pull_request: Add pagination support to `ReferencesClient`

ryangribble commented:
Thanks @gdziadkiewicz !!!

![](https://media2.giphy.com/media/l4FGFoN87zy6DygJW/giphy.gif?cid=e1bb72ff5a21375965354935677d1eea)
